### ISSUE TITLE
WebAssembly driver interface

### ISSUE THREAD

**Issue by localvar** [megaease/easegress#118]
## Background

We have evaluated different WebAssembly engines like [wasmer](https://github.com/wasmerio/wasmer-go), [wasmtime](https://github.com/WasmEdge/WasmEdge), [WasmEdge](https://github.com/WasmEdge/WasmEdge), [aWsm](https://github.com/gwsystems/aWsm) and etc during the preparation for #1 , and we choose wasmtime at last because it is the only one which supports all the required features. However, we think wasmtime is an acceptable choice than the best choice because we observed [virtual memory leaks](https://github.com/bytecodealliance/wasmtime-go/issues/84), and it requires Cgo (Cgo is a common issue of Wasm engines).

On the other hand, as a new and rapid evolution technology, features implemented by Wasm engines also vary from each other. Although Easegress as a Wasm host only requires a limited set of Wasm features, it should provide the maximum flexibility for users to use any Wasm features.

So our target is that Easegress can use different Wasm engines.

## Proposal

We define an interface between Easegress and Wasm engines, and the implementations of the interface are Wasm drivers. By switching between different drivers, Easegress and user Wasm code can adapt different Wasm engines without any changes. Because it much easier for developers to implement an interface than write code for Easegress (which requires developers to understand the Easegress architecture), the drivers can be developed by MegaEase, Wasm engine vendor, or 3rd party developers.

Basically, the driver interface should include the following features:

- create a Wasm VM
- destroy a Wasm VM
- Host call Wasm functions
- Wasm call host functions
- Interrupt Wasm execution

All Wasm engines already support the first 4 items while differ much in implementation details, it is the drivers' responsibility to convert them to the unified interface. For the last one, we only known wasmtime has the support currently (that's why we choose it for #1 ), other Wasm engines must implement it before the driver can be developed.

**Comment by xxx7xxxx** [comment]
Already implemented
